topic/machine-learning/automl/automl_timeseries_forecasting_with_pycaret.ipynb

```diff
@@ -253,7 +289,7 @@
     "data[\"month\"] = pd.to_datetime(data['month'], unit='ms')\n",
     "\n",
     "# We set the MLFLOW_TRACKING_URI to our CrateDB instance. We'll see later why\n",
-    "os.environ[\"MLFLOW_TRACKING_URI\"] = f\"{dburi}&schema=mlflow\""
+    "os.environ[\"MLFLOW_TRACKING_URI\"] = f\"{CONNECTION_STRING}&schema=mlflow\""
```

Comment:
Previously, `dburi` contained an SQLAlchemy database address URI / connection string without the `schema` parameter, so it was good to append one.

Now that it is coming from the `CRATEDB_CONNECTION_STRING` environment variable, which already includes a `schema` parameter, things most probably go south when trying to address the correct database location.

Thanks for the guidance into this direction, @andnig. I think it is a bit sad that the process runs into a timeout situation. While I did not yet look into the details why exactly this is happening, a more explicit exception raised by the software would have been nice, for better UX.

_Note that this issue is exclusively related to software testing, and does not mean the notebook is incorrect. It is only about that it currently doesn't harmonize with automated testing._

Reply:
## Observation

Now that we know what to look for, it is easy to spot the correct exception. It does not help too much that it is printed in green, and that I am not too familiar with analyzing pytest-notebook stacktraces yet.

```
2023/12/03 00:03:58 WARNING mlflow.store.db.utils: SQLAlchemy engine could not be created. The following exception is caught.
(crate.client.exceptions.ProgrammingError) expected string or bytes-like object, got 'tuple'
(Background on this error at: https://sqlalche.me/e/20/f405)
Operation will be retried in 0.1 seconds
```

-- https://github.com/crate/cratedb-examples/actions/runs/7072661550/job/19253059998?pr=174#step:6:952

![image](https://github.com/crate/cratedb-examples/assets/453543/1c082b99-0925-498a-bad1-cf2285642cd0)

## Evaluation

That exception exactly reflects what's happening: The machinery receives **two** values for the `schema` parameter, in form of a tuple, and trips correspondingly.

Reply:
Apparently, there are no time-based adjustments for the reconnection loop of MLflow.

The maximum number of times to retry is configured as `MAX_RETRY_COUNT = 15`.
-- https://github.com/mlflow/mlflow/blob/v2.8.1/mlflow/store/db/utils.py#L55

The reconnection loop will retry for up to 55 minutes (3276.7 seconds), if I got it right.
-- https://github.com/mlflow/mlflow/blob/v2.8.1/mlflow/store/db/utils.py#L224-L243

Reply:
Concluding all of the observations shared above, I think it will be a good idea to probe the database connection before going into the reconnect loop.

-- https://github.com/crate-workbench/mlflow-cratedb

Reply:
Well, this was successful. After fixing the database connectivity settings propagation, test cases signal "green" status now, and there are three more tasks to work on, in order to improve upstream library and adapter software.

- https://github.com/crate/crate-python/issues/598
- https://github.com/crate/crate-python/issues/599
- https://github.com/crate-workbench/mlflow-cratedb/issues/73